### DISCUSSION TITLE
Help creating a multiband compressor

### DISCUSSION THREAD

**Discussion by juanamm** [savonet/liquidsoap#3384] [Q&A]
I'm new and I don't know how to create a multiband compressor for my playlist.
If someone knows how to do it, please adapt it to my code and if you can, explain to me how it works so I can modify it.

```
# Settings
# settings.server.telnet := true
# settings.server.telnet.port := 8595
settings.log.level := 3
settings.harbor.bind_addrs := ["0.0.0.0"]
settings.log.file := true
settings.log.file.path := "/home/icecast2/liquidsoap/log/registro.log"

# SSL certificate configuration
transport = http.transport.tls(
  # Server mode: required,
  # client mode: optional, add certificate to trusted pool
  certificate="/etc/letsencrypt/live/osirisfm.org/fullchain.pem",
  # Server mode: required, client mode: ignored
  key="/etc/letsencrypt/live/osirisfm.org/privkey.pem",
)

# Environment Variables
password_dj=environment.get("PASSWORD_DJ")
password_inv1=environment.get("PASSWORD_INV1")
password_inv2=environment.get("PASSWORD_INV2")
password_inv3=environment.get("PASSWORD_INV3")
source_passw=environment.get("SOURCE_PASSW")
port_https=environment.get("PORT_HTTPS")
port_live=environment.get("PORT_LIVE")
port_inv1=environment.get("PORT_INV1")
port_inv2=environment.get("PORT_INV2")
port_inv3=environment.get("PORT_INV3")
host_https_no_prot=environment.get("HOST_HTTPS_NO_PROT")

# Convert string to integer
port_live_int=int_of_string(port_live)
port_https_int=int_of_string(port_https)
port_inv1_int=int_of_string(port_inv1)
port_inv2_int=int_of_string(port_inv2)
port_inv3_int=int_of_string(port_inv3)

# Set playlist original source and apply nrj
original_playlist_source=playlist("~/music/playlist.pls")
original_playlist=nrj(original_playlist_source)

# Apply crossfades to the playlist
original_playlist=crossfade(fade_out=3.0, fade_in=3.0, original_playlist)

#Here mutliband compressor code ?


# Security fallback
security_source=single("~/music_security/security.mp3")
playlist_securitized=fallback(track_sensitive=false, [original_playlist, security_source])

# Support for live streams
live=input.harbor(
    "live",
    port=port_live_int,
    password=password_dj,
    replay_metadata=true,
    icy_metadata_charset="UTF-8"
)
inv1=input.harbor(
    "inv1",
    port=port_inv1_int,
    password=password_inv1,
    replay_metadata=true,
    icy_metadata_charset="UTF-8"
)
inv2=input.harbor(
    "inv2",
    port=port_inv2_int,
    password=password_inv2,
    replay_metadata=true,
    icy_metadata_charset="UTF-8"
)
inv3=input.harbor(
    "inv3",
    port=port_inv3_int,
    password=password_inv3,
    replay_metadata=true,
    icy_metadata_charset="UTF-8"
)

# Fallback for live and music playlists
radio=fallback(track_sensitive=false, [live, inv1, inv2, inv3, playlist_securitized])  # Usar la fuente combinada

# Stream it out MP3
output.icecast(
    %ffmpeg(format="mp3", %audio(codec="libmp3lame", b="192k")),
    radio,
    mount="stream",
    name="DJ OsiriS Music Service",
    description="DJ OsiriS FM, estamos las 24 horas online con una variada propuesta musical y DJs en vivo!",
    fallible=true,
    host=host_https_no_prot,
    port=port_https_int,
    password=source_passw,
    transport=transport,
)
```

**Comment by vitoyucepi** [comment]
Hi @juanamm,
You could use [`compress.multiband.interactive`](https://www.liquidsoap.info/doc-2.2.0/reference-extras.html#compress.multiband.interactive) or [`compress.multiband`](https://www.liquidsoap.info/doc-2.2.0/reference-extras.html#compress.multiband).
The additional documentation can be found on [page 133 of the book](https://www.liquidsoap.info/book/book.pdf).

**Comment by juanamm** [comment]
> Hi @juanamm, You could use [`compress.multiband.interactive`](https://www.liquidsoap.info/doc-2.2.0/reference-extras.html#compress.multiband.interactive) or [`compress.multiband`](https://www.liquidsoap.info/doc-2.2.0/reference-extras.html#compress.multiband). The additional documentation can be found on [page 133 of the book](https://www.liquidsoap.info/book/book.pdf).

Hi @vitoyucepi , thank you for sharing the necessary links with the syntax and parameters to create the multiband compressor.
The thing is that I'm new, I have a basic level of English and I don't understand how to adapt that code to my script.
I do not want to abuse your kindness, but I ask if you can help me by providing a basic configuration that suits my script and then I can read to change the parameters of the compressor to my liking.
_I only want to apply the compressor to my playlist_, not to the `input.harbor`

I've seen some people have had issues with the metadata and since you helped me migrate to v2.2.0 and everything works great for me, I don't want to mess up and ruin your old work.

It can be possible?

**Reply by vitoyucepi** [reply]
Have you tried using the `interactive.harbor`? Here is some information https://www.liquidsoap.info/doc-dev/server.html#interactive-variables.

I suggest using the compressor like this.
```ruby
#Interactive multiband compressor
original_playlist = compress.multiband.interactive(original_playlist)
```

If this compressor does not meet your needs, there are other options available.

**Comment by juanamm** [comment]
> Have you tried using the `interactive.harbor`? Here is some information https://www.liquidsoap.info/doc-dev/server.html#interactive-variables.
> 
> I suggest using the compressor like this.
> 
> ```ruby
> #Interactive multiband compressor
> original_playlist = compress.multiband.interactive(original_playlist)
> ```
> 
> If this compressor does not meet your needs, there are other options available.

I have correctly configured the code as suggested and added the line to allow to access variables's values ​ by web and that values ​​remain persistent:
```
# Interactive multiband compressor
original_playlist = compress.multiband.interactive(original_playlist)

# Start the web interface
interactive.harbor()

# Save variables values persistent
interactive.persistent("~/liquidsoap/vars.json")
```

Questions:
1.) Can I use some authentication method on the url: 'http://my_host.com:8000/interactive
1.1) Could it be at the level of the liquidosap script, in my case 'main.liq' ?
1.2) Or should I use authentication through 'nginx' config?
Could you share the sample codes?
